tools/dev-doctor/rootcmd.go

```diff
@@ -0,0 +1,161 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"text/tabwriter"
+
+	"github.com/blang/semver"
+	"github.com/spf13/cobra"
+)
+
+var rootCmd = &cobra.Command{
+	Args:  cobra.NoArgs,
+	Short: "Check development setup",
+	Run:   rootCmdRun,
+}
+
+var nfsFirewallChecks *bool
+
+func init() {
+	flags := rootCmd.Flags()
+	nfsFirewallChecks = flags.Bool("nfs-firewall", false, "run extra NFS firewall checks, requires root privileges")
+}
+
+func rootCmdRun(cmd *cobra.Command, args []string) {
+	minGoVersion, err := semver.ParseTolerant(minGoVersionStr)
+	if err != nil {
+		panic(fmt.Sprintf("cannot parse minGoVersionStr: %q", minGoVersionStr))
+	}
+
+	checks := []check{
+		osArchCheck{},
+		unameCheck{},
+		&binaryCheck{
+			name:          "go",
+			ifNotFound:    checkError,
+			versionArgs:   []string{"version"},
+			versionRegexp: regexp.MustCompile(`go version go(\d+\.\d+\S*)`),
+			minVersion:    &minGoVersion,
+		},
+		&binaryCheck{
+			name:          "clang",
+			ifNotFound:    checkError,
+			versionArgs:   []string{"--version"},
+			versionRegexp: regexp.MustCompile(`clang version (\d+\.\d+\.\d+)`),
+			minVersion:    &semver.Version{Major: 10, Minor: 0, Patch: 0},
+		},
+		// FIXME add llvm check?
+		// FIXME add libelf-devel check?
+		&binaryCheck{
+			name:          "ginkgo",
+			ifNotFound:    checkWarning,
+			versionArgs:   []string{"version"},
+			versionRegexp: regexp.MustCompile(`Ginkgo Version (\d+\.\d+\S*)`),
+			minVersion:    &semver.Version{Major: 1, Minor: 4, Patch: 0},
+			hint:          `Run "go get -u github.com/onsi/ginkgo/ginkgo".`,
+		},
+		// FIXME add gomega check?
+		&binaryCheck{
+			name:          "golangci-lint",
+			ifNotFound:    checkWarning,
+			versionArgs:   []string{"version"},
+			versionRegexp: regexp.MustCompile(`(\d+\.\d+\S*)`),
+			minVersion:    &semver.Version{Major: 1, Minor: 27, Patch: 0},
+			hint:          "See https://golangci-lint.run/usage/install/#local-installation.",
+		},
+		&binaryCheck{
+			name:       "docker",
+			ifNotFound: checkError,
+		},
+		&binaryCheck{
+			name:       "docker-compose",
+			ifNotFound: checkWarning,
+		},
+		&binaryCheck{
+			name:          "vagrant",
+			ifNotFound:    checkWarning,
+			versionArgs:   []string{"--version"},
+			versionRegexp: regexp.MustCompile(`Vagrant (\d+\.\d+\S*)`),
+			minVersion:    &semver.Version{Major: 2, Minor: 0, Patch: 0},
+		},
+		&binaryCheck{
+			name:       "virtualbox",
```

Comment:
Can we make that check case insensitive? Currently I'm getting a warning despite the fact, that I have VirtualBox installed as `/usr/bin/VirtualBox`:

```
❯ make dev-doctor
go run ./tools/dev-doctor
RESULT    CHECK            MESSAGE
ok        os/arch          linux/amd64
ok        uname            Linux new-host-6 5.9.0-rc5-1-default+ #2 SMP Mon Oct 26 23:41:40 CET 2020 x86_64 x86_64 x86_64 GNU/Linux
ok        go               found /usr/bin/go, version 1.15.3
ok        clang            found /home/vadorovsky/repos/llvm-project/llvm/build/bin/clang, version 12.0.0
ok        ginkgo           found /home/vadorovsky/go/bin/ginkgo, version 1.14.1
ok        golangci-lint    found /home/vadorovsky/go/bin/golangci-lint, version 1.31.0
ok        docker           found /usr/bin/docker
warning   docker-compose   docker-compose not found in $PATH
ok        vagrant          found /usr/bin/vagrant, version 2.2.10
warning   virtualbox       virtualbox not found in $PATH
ok        pip3             found /usr/bin/pip3, version 20.0.2
ok        docker-group     user vadorovsky in docker group

See https://docs.cilium.io/en/latest/contributing/development/dev_setup/
❯ which VirtualBox
/usr/bin/VirtualBox
❯ which virtualbox
which: no virtualbox in (/home/vadorovsky/repos/llvm-project/llvm/build/bin:/home/vadorovsky/opt/cross/bin:/home/vadorovsky/repos/llvm-project/llvm/build/bin:/home/vadorovsky/opt/cross/bin:/home/vadorovsky/bin:/usr/local/bin:/usr/bin:/bin:/home/vadorovsky/.cargo/bin:/home/vadorovsky/go/bin:/home/vadorovsky/go/src/k8s.io/kubernetes/third_party/etcd:/home/vadorovsky/.npm_packages/bin:/home/vadorovsky/.cargo/bin:/home/vadorovsky/go/bin:/home/vadorovsky/go/src/k8s.io/kubernetes/third_party/etcd:/home/vadorovsky/.npm_packages/bin)

Reply:
This is now done, although with a special case for allowing the `virtualbox` binary to be called `VirtualBox`. It's easy to add more alternative names if needed.

Doing true case sensitivity is a little tricky because you'd either have to call `exec.LookPath` for all 512 possible combinations (e.g. `vIrTuAlBoX`, `ViRtUaLbOx`, etc.) or write code to read all directories in `$PATH` and check for binaries with matching names.

Reply:
You are right. Thanks for the fix!